ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala

```diff
@@ -655,22 +655,54 @@ private class JdbcLedgerDao(
     logger.info("Storing package entry")
     dbDispatcher.executeSql(metrics.daml.index.db.storePackageEntryDbMetrics) {
       implicit connection =>
-        val update = optEntry.map {
-          case PackageLedgerEntry.PackageUploadAccepted(submissionId, recordTime) =>
-            Update.PublicPackageUpload(
-              archives = packages.view.map(_._1).toList,
-              sourceDescription = packages.headOption.flatMap(
-                _._2.sourceDescription
-              ), // this is valid since all clients of this method share the same behavior: all PackageDetails-s will have the same sourceDescription
-              recordTime = Time.Timestamp.assertFromInstant(recordTime),
-              submissionId = Some(submissionId),
+        // Note on knownSince and recordTime:
+        // - There are two different time values in the input: PackageDetails.knownSince and PackageUploadAccepted.recordTime
+        // - There is only one time value in the intermediate values: PublicPackageUpload.recordTime
+        // - There are two different time values in the database schema: packages.known_since and package_entries.recorded_at
+        // This is not an issue since all callers of this method always use the same value for all knownSince and recordTime times.
+        //
+        // Note on sourceDescription:
+        // - In the input, each package can have its own source description (see PackageDetails.sourceDescription)
+        // - In the intermediate value, there is only one source description for all packages (see PublicPackageUpload.sourceDescription)
+        // - In the database schema, each package can have its own source description (see packages.source_description)
+        // This is again not an issue since all callers of this method always use the same value for all source descriptions.
+        val update = optEntry match {
+          case None =>
+            if (packages.nonEmpty) {
```

Comment:
this if condition is not needed: in UpdateToDBDTOV1 there will be an empty iterator produced as soon as submissionId is empty and packages is empty.

Reply:
also if the if condition is dropped all the updates will be Some(...) so we can drop the Some(...) from everywhere and do it when we use the update in sequentialIndexer.store

Reply:
> this if condition is not needed

It's needed for the code below, because it uses `packages.head`, and I didn't want to rely too much on the fact that the packages will never be empty. I'll try to rewrite this whole block.

Reply:
oh yeah, right!